File: 40-Combination Sum II-组合总和 II/40-Combination Sum II-组合总和 II.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */

 let res = []
 let path = []
 let sum = 0
 
 var combinationSum2 = function(candidates, target) {
     res = []
     candidates.sort((a, b) => a-b)
     backtradce(candidates, target, 0)
     return res
 };
 
 let backtradce = function(candidates, target, index){
     if(sum > target) return
     if(sum === target){
         res.push(Array.from(path))
         return
     }
 
     
     for(let i = index; i < candidates.length; i++){
         if(i > index && candidates[i] == candidates[i-1]) continue;
         path.push(candidates[i])
         sum += candidates[i]
         backtradce(candidates, target, i + 1)
         path.pop()
         sum -= candidates[i]
     }
 }

File: 40-Combination Sum II-组合总和 II/40-Combination Sum II-组合总和 II.java
class Solution {
    public List<List<Integer>> res;
    public int sum;
    public LinkedList<Integer> path;

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        res = new ArrayList<>();
        path = new LinkedList<Integer>();
        Arrays.sort( candidates );
        backtrace(candidates, target, 0);
        return res;
    }

    public void backtrace(int[] candidates, int target, int index){
        if(sum > target){
            return;
        }
        if(sum == target){
            res.add(new ArrayList<Integer>(){{addAll(path);}});
            return;
        }

        for(int i = index; i < candidates.length; i++){
            if(target<candidates[i]) break;
            if(i > index && candidates[i] == candidates[i - 1]) continue;
            path.add(candidates[i]);
            sum += candidates[i];
            backtrace(candidates, target, i + 1);
            path.removeLast();
            sum -= candidates[i];
        }
    }
}